drv/lpc55-update-server/src/images.rs

```diff
@@ -26,79 +47,617 @@ extern "C" {
     // for either position to be used in new images, we clear bit 28 in all of
     // the numbers used for comparison below, by ANDing them with this mask:
-    const ADDRMASK: u32 = !(1 << 28);
+    pub fn normalized_initial_pc(&self) -> u32 {
+        const ADDRMASK: u32 = !(1 << 28);
+        self.initial_pc & ADDRMASK
+    }
+
+    // Length of image from offset zero to end of the signature block (without padding)
+    // a.k.a. ImageVectorsLpc55.nxp_image_length
+    pub fn image_length(&self) -> Option<u32> {
+        if self.is_image_type_signed_xip() {
+            Some(self.nxp_image_length)
+        } else {
+            None
+        }
+    }
 
-    let reset_vector = u32::from_le_bytes(
-        block[RESET_VECTOR_OFFSET..][..4].try_into().unwrap_lite(),
-    ) & ADDRMASK;
+    /// Image length padded to nearest page size.
+    pub fn padded_image_len(&self) -> Option<u32> {
+        round_up_to_flash_page(self.image_length()?)
+    }
 
-    // Ensure the image is destined for the right target
-    if !exec.contains(&reset_vector) {
+    /// Determine the bounds of an image assuming the given flash bank
+    /// addresses.
+    pub fn padded_image_range(
+        &self,
+        at_runtime: &Range<u32>,
+    ) -> Option<Range<u32>> {
+        let image_start = at_runtime.start;
+        let image_end =
+            at_runtime.start.checked_add(self.padded_image_len()?)?;
+        Some(image_start..image_end)
+    }
+}
+
+impl TryFrom<&[u8]> for ImageVectorsLpc55 {
+    type Error = ();
+
+    fn try_from(buffer: &[u8]) -> Result<Self, Self::Error> {
+        match ImageVectorsLpc55::read_from_prefix(buffer) {
+            Some(vectors) => Ok(vectors),
+            None => Err(()),
+        }
+    }
+}
+
+/// Sanity check the image header block.
+/// Return the offset to the end of the executable code which is also
+/// the end of optional caboose and the beginning of the signature block.
+pub fn validate_header_block(
+    header_access: &ImageAccess<'_>,
+) -> Result<u32, UpdateError> {
+    let mut vectors = ImageVectorsLpc55::new_zeroed();
+    let mut header = ImageHeader::new_zeroed();
+
+    // Read block 0 and the header contained within (if available).
+    if header_access.read_bytes(0, vectors.as_bytes_mut()).is_err()
+        || header_access
+            .read_bytes(IMAGE_HEADER_OFFSET, header.as_bytes_mut())
+            .is_err()
+    {
+        return Err(UpdateError::InvalidHeaderBlock);
+    }
+
+    // Check image type and presence of signature block.
+    if !vectors.is_image_type_signed_xip()
+        || vectors.nxp_offset_to_specific_header >= vectors.nxp_image_length
+    {
+        // Not a signed XIP image or no signature block.
+        // If we figure out a reasonable minimum size for the signature block
+        // we should test for that.
+        return Err(UpdateError::InvalidHeaderBlock);
+    }
+
+    // We don't rely on the ImageHeader, but if it is there, it needs to be valid.
```

Comment:
This PR doesn't remove the `ImageHeader`, right?

I'm wondering why the logic change here; previously, we required the `ImageHeader` if the component is `RotComponent::Hubris` (but allowed for _bootloaders_ that have been released without an `ImageHeader`).